Refactor CEH model

thfroitzheim opened the pull request:
The CEH model implemented in #124 was refactored to follow the layout for the GFN Hamiltonians more closely. This includes: 
1. Transfer of ceh_guess into its own singlepoint.f90 file.
2. Transfer of the Hamiltonian setup and the (rewritten) get_hamiltonian function into the h0.f90 file. In this course, the new ceh_hamiltonian and ceh_calculator classes were removed and replaced by the tb_hamiltonian and xtb_calculator base classes (added data will also be required in the GP3-xTB development). 
3. Introduction of a ceh_h0spec to use the tb_hamiltonian machinery for the Hamiltonian setup. Expansion of the getter functions needed for get_hamiltonian. 
4. Consistent renaming of all variables to match the prior GFN implementation. 
5. Expansion of the CEH-specific test suite. 
The restructuring is necessary to use the already available GFN implementation for the setup of analytical charge gradients of CEH. 

The coordination number was restructured to prevent code duplication due to the many different CNs (now renamed). A new CN now only requires defining the count function and its derivative. To allow directed CN functions (i.e. erf_en, added to one partner and subtracted from the other) a directed_factor is included (I am open to better ideas). In this course, the testing was expanded to cover all possible CN implementations. 

Lastly, the derivatives of the diatomic frame scaled overlap was implemented.

awvwgk commented on src/tblite/ceh/singlepoint.f90:
```diff
@@ -0,0 +1,251 @@
+      type(timer_type) :: timer
+      real(wp) :: ttime
+
+      logical :: grad = .false.
```
This variable is implicitly saved, this is a bug.

thfroitzheim replied:
That is good to know. I removed all cases, where it happend

awvwgk commented on src/tblite/ceh/singlepoint.f90:
```diff
@@ -0,0 +1,251 @@
+      logical :: grad = .false.
+
+      real(wp) :: elec_entropy
+      real(wp) :: nel = 0.0_wp, cutoff
```
Also, this variable is implicitly saved, this is generally not the wanted behavior.

awvwgk commented on src/tblite/ceh/singlepoint.f90:
```diff
@@ -0,0 +1,251 @@
+      call timer%pop
+      ttime = timer%get("wall time CEH")
+
+      !> Printout of results
+      if (prlevel > 2) then
+         call ctx%message(label_charges)
+         call ctx%message("Atom index      Charge / a.u.")
+         do i = 1, mol%nat
+            call ctx%message(format_string(i, "(i7)") // &
+            & "   " // format_string(wfn%qat(i,1), real_format))
+         end do
+         call ctx%message(repeat("-", 60))
+         call ctx%message(label_dipole)
+         call ctx%message("     x           y           z")
+         call ctx%message(format_string(dipole(1), "(f12.5)") // &
+         & format_string(dipole(2), "(f12.5)") // &
+         & format_string(dipole(3), "(f12.5)"))
+         call ctx%message(repeat("-", 60))
+      endif
```
Is this printout necessary here or can it be implemented on a higher level, e.g. in the guess or main driver?

thfroitzheim replied:
No that is not needed. I remove it.

awvwgk commented on src/tblite/ncoord/erf.f90:
```diff
@@ -0,0 +1,137 @@
+
+      real(wp) :: rc, exponent, expterm, count
+
+      rc = self%rcov(izp) + self%rcov(jzp)
+
+      exponent = kcn*(r-rc)/rc
+
+      expterm = exp(-exponent**2)
+
+      count = -(kcn*expterm)/(rc*sqrt(pi))
+
+
+   end function ncoord_dcount
+
+
```
Please check whether the extra empty lines are necessary here.

awvwgk commented on src/tblite/ncoord/exp.f90:
```diff
@@ -18,73 +18,121 @@
    use mctc_env, only : wp
    use mctc_io, only : structure_type
-   use tblite_cutoff, only : get_lattice_points
+   !use tblite_cutoff, only : get_lattice_points
```
Please removed commented out code.

awvwgk reviewed (changes requested):
Thanks for taking the effort of merging the CEH hamiltonian with the xTB one, this is a quite positive change. I left a few comments on issues and bugs which I could spot while skimming through the PR.

thfroitzheim commented:
I fixed all of the formatting issues you pointed out and removed unnecessary output and the implicitly saved variable definitions. In this course, I updated also the naming of the tests as now many versions of ncoord are tested, which led to a small merge conflict (I think I resolved it correctly).

It took a bit longer because there was already the first update for the CEH parameters. I updated all parameters and moved the radii and EN values into the ceh.f90 file, as they are also fitted. Otherwise, there was no major change to the code.

awvwgk commented on app/cli.f90:
```diff
@@ -429,6 +433,12 @@ subroutine get_run_arguments(config, list, start, error)
          call get_argument_as_real(arg, config%etemp, error)
          if (allocated(error)) exit
 
+      case("--etemp_guess")
```
Let's use dashes rather than underscores.

awvwgk commented on src/tblite/ceh/singlepoint.f90:
```diff
@@ -0,0 +1,207 @@
+
+      integer :: i, prlevel
+
+      !> coordination number related arrays
```
Please don't use docstring markers for non-docstring comments. These comments are prone to end up in the wrong place when extracted by ford or doxygen. Only dummy variables or procedures should have comments with the docstring marker `>`.

thfroitzheim replied:
I have found a few more and removed them.

awvwgk commented on src/tblite/ncoord/exp.f90:
```diff
@@ -18,73 +18,106 @@
 !> Provides a coordination number implementation with exponential counting function
 
 !> Coordination number implementation using an exponential counting function
+!> Same as in dftd3 reimplemented
```
Does this need a dot or comma at the end of the previous sentence?

awvwgk reviewed (approved):
Only some minor comments, now it looks like in a good shape to get merged.

thfroitzheim commented on src/tblite/ceh/h0.f90:
```diff
@@ -16,18 +16,292 @@
+            selfenergy(ii+ish) = h0%selfenergy(ish, izp)
+         end do
+      end do
+      if (present(cn) .and. present(cn_en)) then
```
That is a better way.

thfroitzheim commented:
Great. I have fixed the few points you mentioned and also updated the help page for the guess.

awvwgk commented:
Can you please also make sure the entries in man/ are consistent with the help output?

thfroitzheim commented:
I have added a new page to man for the guess and the guess temperature to the page for run.

awvwgk reviewed (approved):
Looks good to me now

thfroitzheim commented:
I think that you will have to merge the PR as I am not a maintainer (there is no hurry, but you don't have to wait for me).
